### PR TITLE
[For v0.18] Sketcher: External Geometry as defining

### PR DESCRIPTION
**Pull request by abdullahtahiriyo**
https://forum.freecadweb.org/viewtopic.php?f=10&t=27727#p224065

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.

---

### CODE REVIEW

File: src/Mod/Sketcher/App/SketchObjectPyImp.cpp [hunk 1]
```diff
@@ -268,13 +266,63 @@ PyObject* SketchObjectPy::setConstruction(PyObject *args)
...
+}
+
+PyObject* SketchObjectPy::setExternalDefining(PyObject *args)
+{
+    int Index;
+    PyObject *Mode;
+    if (!PyArg_ParseTuple(args, "iO!", &Index, &PyBool_Type, &Mode))
+        return 0;
+
+    if (Index<0 || this->getSketchObjectPtr()->setConstruction(-3-Index, !(PyObject_IsTrue(Mode) ? true : false))) {
```
Comment (wwmayer) [inline src/Mod/Sketcher/App/SketchObjectPyImp.cpp]: Shouldn't -3 be replaced with GeoEnum::RefExt?

**Comment by abdullahtahiriyo** [comment]
@wwmayer 

I am back and I can respond in a reasonable time now. so if you wish we can continue with the review. This is what has been done, apart from changing the -3, -2, -1 indices to the corresponding GeoEnums:
https://forum.freecadweb.org/viewtopic.php?f=10&t=27780&start=10#p226963

**Comment by wwmayer** [comment]
I will try to test your branch soon and give you some more feedback.

**Comment by abdullahtahiriyo** [comment]
In this last commit I have heavily refactored PropertyLinkSubList, creating several protected functions to allow to maintain the members as private. Part of the refactoring looks very good to me and was long due (for example the maintenance of backlinks). Other parts look more or less good, like the assigning functions that allow the derived class to reuse more code. Some other parts, like the access from removeValue are not very elegant, and basically by-pass the fact that the members are private by using a getValue(int index) kind of function... I am sure you will have some further comments :) Thanks for taking the time!

**Comment by abdullahtahiriyo** [comment]
This branch is in conflict. I close this PR in favour of this other that is rebased and without conflict:
https://github.com/FreeCAD/FreeCAD/pull/1421

**Pull request closed without merging**